ams/ams-server/src/main/java/com/netease/arctic/ams/server/service/impl/OptimizeQueueService.java

```diff
@@ -484,117 +484,137 @@ public String queueName() {
 
     public OptimizeTask poll(JobId jobId, final String attemptId, long waitTime) {
       long startTime = System.currentTimeMillis();
+      OptimizeTaskItem task = tasks.poll();
+      if (task == null) {
+        tryPlanAsync(jobId, attemptId);
+        task = waitForTask(startTime, waitTime);
+        if (task == null) {
+          return null;
+        } else {
+          return onExecuteOptimizeTask(task, jobId, attemptId);
+        }
+      } else {
+        return onExecuteOptimizeTask(task, jobId, attemptId);
+      }
+    }
+
+    private OptimizeTask onExecuteOptimizeTask(OptimizeTaskItem task, JobId jobId, String attemptId) {
+      TableTaskHistory tableTaskHistory;
+      try {
+        // load files from sysdb
+        task.setFiles();
+        // update max execute time
+        task.setMaxExecuteTime();
+        tableTaskHistory = task.onExecuting(jobId, attemptId);
+      } catch (Exception e) {
+        task.clearFiles();
+        LOG.error("{} handle sysdb failed, try put task back into queue", task.getTaskId(), e);
+        if (!tasks.offer(task)) {
+          LOG.error("{} failed to put task back into queue", task.getTaskId());
+          task.onFailed(new ErrorMessage(System.currentTimeMillis(), "failed to put task back into queue"), 0);
+        }
+        throw e;
+      }
+      try {
+        insertTableTaskHistory(tableTaskHistory);
+      } catch (Exception e) {
+        LOG.error("failed to insert tableTaskHistory, {} ignore", tableTaskHistory, e);
+      }
+      return task.getOptimizeTask();
+    }
+
+    private OptimizeTaskItem waitForTask(long startTime, long waitTime) {
       while (true) {
+        lock();
+        try {
+          // if timeout, return null, await support zero or a negative value, and return false
+          if (planThreadCondition.await(waitTime - (System.currentTimeMillis() - startTime),
+              TimeUnit.MILLISECONDS)) {
+            OptimizeTaskItem task = tasks.poll();
+            if (task != null) {
+              if (!tables.contains(task.getTableIdentifier())) {
```

Comment:
It seems that we not only need to check the legality of the task here, but also the task obtained at the beginning.

Reply:
I created a method pollValidTask() to check legality while polling tasks.
```
    private OptimizeTaskItem pollValidTask() {
      while (true) {
        OptimizeTaskItem task = tasks.poll();
        if (task == null) {
          return null;
        }
        if (tables.contains(task.getTableIdentifier())) {
          return task;
        } else {
          LOG.warn("get task {} from queue {} but table {} not in this queue, continue",
              task.getTaskId(), queueName(), task.getTableIdentifier());
        }
      }
    }
```